File: ffi/ts/src/tokenId.ts
import {
  castToTokenId,
  deserializeTokenId,
  genDefaultTokenId,
  genTokenId,
  genTokenIdWithSubid,
  getTokenIdSubid,
  getTokenIdToken,
  freeObj,
  serializeTokenId,
} from './blsct'

import { ManagedObj } from './managedObj'

/** Represents a token ID. A token ID consists of two parameters: token and subid, both of which are optional. When omitted, default values are used instead of random values.
 * 
 * Examples:
 * ```ts
 * const { TokenId } = require('navio-blsct')
 * const tid1 = TokenId.default()
 * const tid2 = TokenId.fromToken(123)
 * const tid3 = TokenId.fromTokenAndSubid(123, 456)
 * tid3.getToken() // 123
 * tid3.getSubid() // 456
 * tid3.equals(tid2) // false
 * tid3.equals(tid3) // true
 * ser = tid3.serialize()
 * deser = TokenId.deserialize(ser)
 * deser.serialize() === ser // true
 * ```
 */  
export class TokenId extends ManagedObj {
  constructor(obj: any) {
    super(obj)
  }

  /** Generates a new default `TokenId` instance.
   * @returns A new `TokenId` instance with default values for token and subid.
   */
  static default(): TokenId {
    const rv = genDefaultTokenId()
    return new TokenId(rv.value)
  }

  /** Generates a `TokenId` from the provided token.
   *
   * @param token - The token number to use for the `TokenId`.
   * @return A new `TokenId` instance with the specified token and a default subid.
   */
  static fromToken(token: number): TokenId {
    const rv = genTokenId(token)
    const tokenId = TokenId.fromObj(rv.value)
    freeObj(rv)
    return tokenId
  }

  /** Generates a `TokenId` from the provided token and subid.
   *
   * @param token - The token number to use for the `TokenId`.
   * @param subid - The subid number to use for the `TokenId`.
   * @return A new `TokenId` instance with the specified token and subid.
   */
  static fromTokenAndSubid(token: number, subid: number): TokenId {
    const rv = genTokenIdWithSubid(token, subid)
    const tokenId = TokenId.fromObj(rv.value)
    freeObj(rv)
    return tokenId
  }

  /** Returns the token number of the `TokenId`.
   *
   * @return The token number of the `TokenId`.
   */
  getToken(): number {
    return getTokenIdToken(this.value())
  }

  /** Returns the subid number of the `TokenId`.
   *
   * @return The subid number of the `TokenId`.
   */
  getSubid(): number {
    return getTokenIdSubid(this.value())
  }

  /** Checks if the current `TokenId` is equal to another `TokenId`.
   *
   * @param other - The `TokenId` to compare with.
   * @return `true` if both token and subid are equal, `false` otherwise.
   */
  equals(other: TokenId): boolean {
    return this.getToken() === other.getToken() &&
      this.getSubid() === other.getSubid()
  }

  override value(): any {
    return castToTokenId(this.obj)
  }

  override serialize(): string {
    return serializeTokenId(this.value())
  }

  /** Deserializes a hexadecimal string into a `TokenId` instance.
   *
   * @param hex - The hexadecimal string to convert.
   * @returns The `TokenId` instance represented by the input string.
   */
  static deserialize(
    this: new (obj: any) => TokenId,
    hex: string
  ): TokenId {
    return TokenId._deserialize(hex, deserializeTokenId)
  }
}

File: ffi/rust/src/token_id.rs
use crate::{
  blsct_obj::BlsctObj,
  blsct_serde::BlsctSerde, 
  ffi::{
    BlsctRetVal,
    BlsctTokenId,
    deserialize_token_id,
    gen_default_token_id,
    gen_token_id,
    get_token_id_subid,
    get_token_id_token,
    gen_token_id_with_token_and_subid,
    serialize_token_id,
  },
  macros::{
    impl_clone,
    impl_display,
    impl_from_retval,
    impl_value,
  },
};
use std::ffi::c_char;
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize, Serialize, Eq)]
pub struct TokenId {
  obj: BlsctObj<TokenId, BlsctTokenId>,
}

impl_from_retval!(TokenId);
impl_display!(TokenId);
impl_clone!(TokenId);

impl TokenId {
  pub fn default() -> Self {
    let rv = unsafe { gen_default_token_id() };
    BlsctObj::from_retval(rv).unwrap().into()
  }

  pub fn from_token(token: u64) -> Self {
    let rv = unsafe { gen_token_id(token) };
    BlsctObj::from_retval(rv).unwrap().into()
  }

  pub fn from_token_and_subid(token: u64, subid: u64) -> Self {
    let rv = unsafe { gen_token_id_with_token_and_subid(token, subid) };
    BlsctObj::from_retval(rv).unwrap().into()
  }

  pub fn token(&self) -> u64 {
    unsafe { get_token_id_token(self.value()) }
  }

  pub fn subid(&self) -> u64 {
    unsafe { get_token_id_subid(self.value()) }
  }

  impl_value!(TokenId, BlsctTokenId);
}

impl BlsctSerde for TokenId {
  unsafe fn serialize(ptr: *const u8) -> *const i8 {
    serialize_token_id(ptr as *const BlsctTokenId)
  }

  unsafe fn deserialize(hex: *const c_char) -> *mut BlsctRetVal {
    unsafe { deserialize_token_id(hex) }
  }
}

impl From<BlsctObj<TokenId, BlsctTokenId>> for TokenId {
  fn from(obj: BlsctObj<TokenId, BlsctTokenId>) -> TokenId {
    TokenId { obj }
  }
}

impl PartialEq for TokenId {
  fn eq(&self, other: &Self) -> bool {
    self.token() == other.token() && self.subid() == other.subid()
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use crate::initializer::init;

  #[test]
  fn test_default() {
    init();
    let token_id = TokenId::default();
    assert_eq!(token_id.token(), 0);
    assert_eq!(token_id.subid(), u64::MAX); // should be uint64 max
  }

  #[test]
  fn test_from_token() {
    init();
    let token = 123u64;
    let token_id = TokenId::from_token(token);
    assert_eq!(token_id.token(), token);
    assert_eq!(token_id.subid(), u64::MAX); // should be uint64 max
  }

  #[test]
  fn test_from_token_and_subid() {
    init();
    let token = 123u64;
    let subid = 456u64;
    let token_id = TokenId::from_token_and_subid(token, subid);
    assert_eq!(token_id.token(), token);
    assert_eq!(token_id.subid(), subid);
  }

  #[test]
  fn test_deser() {
    init();

    let a = TokenId::from_token_and_subid(123, 456);
    let hex = bincode::serialize(&a).unwrap();
    let b = bincode::deserialize::<TokenId>(&hex).unwrap();
    assert!(a == b);
  }
}

